File: tools/UmbraManagerWpf/UmbraManager/Services/MetricsStore.cs
using Microsoft.Data.Sqlite;

namespace UmbraManager.Services;

public sealed class MetricsStore : IDisposable
{
    private readonly SqliteConnection _conn;

    public MetricsStore()
    {
        var dir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "UmbraManager");
        Directory.CreateDirectory(dir);
        _conn = new SqliteConnection($"Data Source={Path.Combine(dir, "manager.db")}");
        _conn.Open();
        EnsureSchema();
    }

    private void EnsureSchema()
    {
        using var cmd = _conn.CreateCommand();
        cmd.CommandText = """
            CREATE TABLE IF NOT EXISTS metrics (
              id INTEGER PRIMARY KEY AUTOINCREMENT,
              service_id TEXT NOT NULL,
              ts INTEGER NOT NULL,
              cpu_pct REAL,
              mem_mb REAL,
              uptime_s INTEGER);
            CREATE INDEX IF NOT EXISTS idx_metrics_service_ts ON metrics(service_id, ts);
            """;
        cmd.ExecuteNonQuery();
    }

    public void RecordMetric(string serviceId, double cpuPct, double memMb, int uptimeS)
    {
        using var cmd = _conn.CreateCommand();
        cmd.CommandText = "INSERT INTO metrics (service_id, ts, cpu_pct, mem_mb, uptime_s) VALUES ($s,$t,$c,$m,$u)";
        cmd.Parameters.AddWithValue("$s", serviceId);
        cmd.Parameters.AddWithValue("$t", DateTimeOffset.UtcNow.ToUnixTimeSeconds());
        cmd.Parameters.AddWithValue("$c", cpuPct);
        cmd.Parameters.AddWithValue("$m", memMb);
        cmd.Parameters.AddWithValue("$u", uptimeS);
        cmd.ExecuteNonQuery();
    }

    public List<(DateTime Time, double Value)> CpuHistory(string serviceId, int minutes = 60)
    {
        var since = DateTimeOffset.UtcNow.AddMinutes(-minutes).ToUnixTimeSeconds();
        var list = new List<(DateTime, double)>();
        using var cmd = _conn.CreateCommand();
        cmd.CommandText = "SELECT ts, cpu_pct FROM metrics WHERE service_id=$s AND ts>=$t ORDER BY ts";
        cmd.Parameters.AddWithValue("$s", serviceId);
        cmd.Parameters.AddWithValue("$t", since);
        using var r = cmd.ExecuteReader();
        while (r.Read())
            list.Add((DateTimeOffset.FromUnixTimeSeconds(r.GetInt64(0)).LocalDateTime, r.GetDouble(1)));
        return list;
    }

    public List<(DateTime Time, double Value)> MemHistory(string serviceId, int minutes = 60)
    {
        var since = DateTimeOffset.UtcNow.AddMinutes(-minutes).ToUnixTimeSeconds();
        var list = new List<(DateTime, double)>();
        using var cmd = _conn.CreateCommand();
        cmd.CommandText = "SELECT ts, mem_mb FROM metrics WHERE service_id=$s AND ts>=$t ORDER BY ts";
        cmd.Parameters.AddWithValue("$s", serviceId);
        cmd.Parameters.AddWithValue("$t", since);
        using var r = cmd.ExecuteReader();
        while (r.Read())
            list.Add((DateTimeOffset.FromUnixTimeSeconds(r.GetInt64(0)).LocalDateTime, r.GetDouble(1)));
        return list;
    }

    public void Dispose() => _conn.Dispose();
}

File: tools/UmbraManager.qt-archive/Core/MetricsStore.cpp
#include "MetricsStore.hpp"
#include <QSqlQuery>
#include <QSqlError>
#include <QStandardPaths>
#include <QDir>

MetricsStore::MetricsStore(QObject* parent) : QObject(parent) {}

MetricsStore::~MetricsStore() {
  if (db_.isOpen()) {
    db_.close();
  }
}

bool MetricsStore::initialize() {
  const QString dir = QStandardPaths::writableLocation(QStandardPaths::AppLocalDataLocation);
  QDir().mkpath(dir);
  const QString dbPath = dir + "/manager.db";

  db_ = QSqlDatabase::addDatabase("QSQLITE", "umbra_metrics");
  db_.setDatabaseName(dbPath);
  if (!db_.open()) return false;
  return ensureSchema();
}

bool MetricsStore::ensureSchema() {
  QSqlQuery q(db_);
  const char* sql =
      "CREATE TABLE IF NOT EXISTS metrics ("
      "id INTEGER PRIMARY KEY AUTOINCREMENT,"
      "service_id TEXT NOT NULL,"
      "ts INTEGER NOT NULL,"
      "cpu_pct REAL,"
      "mem_mb REAL,"
      "uptime_s INTEGER);"
      "CREATE INDEX IF NOT EXISTS idx_metrics_service_ts ON metrics(service_id, ts);";
  return q.exec(sql);
}

void MetricsStore::recordMetric(const QString& serviceId, const QJsonObject& stats) {
  QSqlQuery q(db_);
  q.prepare("INSERT INTO metrics (service_id, ts, cpu_pct, mem_mb, uptime_s) VALUES (?, ?, ?, ?, ?)");
  q.addBindValue(serviceId);
  q.addBindValue(QDateTime::currentSecsSinceEpoch());
  q.addBindValue(stats.value("cpu_pct").toDouble());
  q.addBindValue(stats.value("mem_mb").toDouble());
  q.addBindValue(stats.value("uptime_s").toInt());
  q.exec();
}

QList<QPair<QDateTime, double>> MetricsStore::cpuHistory(const QString& serviceId, int minutes) const {
  QList<QPair<QDateTime, double>> out;
  const qint64 since = QDateTime::currentSecsSinceEpoch() - minutes * 60;
  QSqlQuery q(db_);
  q.prepare("SELECT ts, cpu_pct FROM metrics WHERE service_id=? AND ts>=? ORDER BY ts");
  q.addBindValue(serviceId);
  q.addBindValue(since);
  if (!q.exec()) return out;
  while (q.next()) {
    out.append({QDateTime::fromSecsSinceEpoch(q.value(0).toLongLong()), q.value(1).toDouble()});
  }
  return out;
}

QList<QPair<QDateTime, double>> MetricsStore::memHistory(const QString& serviceId, int minutes) const {
  QList<QPair<QDateTime, double>> out;
  const qint64 since = QDateTime::currentSecsSinceEpoch() - minutes * 60;
  QSqlQuery q(db_);
  q.prepare("SELECT ts, mem_mb FROM metrics WHERE service_id=? AND ts>=? ORDER BY ts");
  q.addBindValue(serviceId);
  q.addBindValue(since);
  if (!q.exec()) return out;
  while (q.next()) {
    out.append({QDateTime::fromSecsSinceEpoch(q.value(0).toLongLong()), q.value(1).toDouble()});
  }
  return out;
}

int MetricsStore::totalPlayersOnline() const { return totalPlayersOnline_; }
void MetricsStore::setTotalPlayersOnline(int count) { totalPlayersOnline_ = count; }
